libs/extensions/src/NanoEvent.cpp

```diff
@@ -106,6 +106,129 @@ shared_ptr<NanoMuons> NanoEvent::GetSegmentMatchedMuons(shared_ptr<NanoMuons> mu
   return allMuons;
 }
 
+shared_ptr<PhysicsObject> NanoEvent::GetSegmentMatchedBestMuonVertex(shared_ptr<PhysicsObject> bestVertex, shared_ptr<PhysicsObjects> goodVerticesCollection, float minMatchRatio) {
+  auto nanoVertex = asNanoDimuonVertex(bestVertex,event);
+  // Pat-PAT dimuon vertex
+  if (nanoVertex->IsPatDimuon()) return bestVertex;
+
+  auto patVertexCollection = make_shared<PhysicsObjects>();
+  auto patDSAVertexCollection = make_shared<PhysicsObjects>();
+  for (auto goodVertex : *goodVerticesCollection) {
+    if (asNanoDimuonVertex(goodVertex,event)->IsPatDimuon()) {
+      patVertexCollection->push_back(goodVertex);
+    }
+    if (asNanoDimuonVertex(goodVertex,event)->IsPatDSADimuon()) {
+      patDSAVertexCollection->push_back(goodVertex);
+    }
+  }
+  // DSA-DSA dimuon vertex
+  if (nanoVertex->IsDSADimuon()) {
+    auto dsaMuon1 = nanoVertex->Muon1();
+    auto dsaMuon2 = nanoVertex->Muon2();
+    float nSegments1 = dsaMuon1->Get("nSegments");
+    float nSegments2 = dsaMuon2->Get("nSegments");
+    vector<int> patMatchIndices1;
+    vector<int> patMatchIndices2;
+    // Get all mathed PAT muons
+    for (int i = 1; i <= 5; i++) {
+      float ratio_tmp1 = dsaMuon1->GetMatchesForNthBestMatch(i) / nSegments1;
+      float ratio_tmp2 = dsaMuon2->GetMatchesForNthBestMatch(i) / nSegments2;
+
+      if(ratio_tmp1 >= minMatchRatio) {
+        patMatchIndices1.push_back(dsaMuon1->GetMatchIdxForNthBestMatch(i));
+      }
+      if(ratio_tmp2 >= minMatchRatio) {
+        patMatchIndices2.push_back(dsaMuon2->GetMatchIdxForNthBestMatch(i));
+      }
+    }
+    int matchedVertexIdx = -1;
+    float minChi2 = 9999.;
+    // Check if any PAT muon combinations are in patVertexCollection
+    // If more than one vertex we select the one with lowest chi2
+    for (auto matchIndex1 : patMatchIndices1) {
+      for (auto matchIndex2 : patMatchIndices2) {
+        if (matchIndex1 == matchIndex2) continue;
+        for (int i=0; i<patVertexCollection->size(); i++) {
+          auto patVertex = patVertexCollection->at(i);
+          auto muon1 = asNanoDimuonVertex(patVertex,event)->Muon1();
+          auto muon2 = asNanoDimuonVertex(patVertex,event)->Muon2();
+          if (((float)muon1->Get("idx") == matchIndex1 && (float)muon2->Get("idx") == matchIndex2) || 
+            ((float)muon1->Get("idx") == matchIndex2 && (float)muon2->Get("idx") == matchIndex1)) {          
```

Comment:
For instance, some of these operations could be hidden inside `NanoMuon`, and then perhaps the whole check could be hidden in `NanoDimuonVertex` to make it a bit more compact here.